rfc/drafts/mediachain-rfc-3.md

```diff
@@ -0,0 +1,449 @@
+- `Journal` - Flattened and partially-reconciled view of the Mediachain knowledge base. Maintained by `Transactors`.
+- `Folding`: Producing a single flattened representation of a chain of updates, reconciling any conflicts and discarding bad information in the process.
+- `Clients` - Embedded API or external nodes, which are the interface for end-user nodes to receive a stream of updates to the `Journal` from `Transactor` nodes, and for communicating writes to the `Transactors`. They also conduct reads and writes to the `Datastore`.
+- `Transactors` - Core of Mediachain. Accepts writes to the `Journal` from `Clients`, serves `Clients` with views of the `Journal`, and records transactions on their behalf. (Formerly known as `Peers`.)
+
+
+## Roadmap
+
+`V1 Indexer` will be a minimal proof of concept, while `Future Indexer` represents a set of future enhancements. Key differences are outlined below:
+
+
+   `V1 Indexer`                                      |  `Future Indexer`
+  ---------------------------------------------------|------------------------------------------------------------------------
+   Centralized.                                      | All components distributed across multiple machines.
+   Off-the-shelf components.                         | Purpose-built components.
+   Single perceptual matching goal.                  | Multiple semantic matching intents supported.
+   No supervised training.                           | Custom supervised relevancy / dedupe training for specific task(s) and dataset at hand.
+   No multi-object detection.                        | Multiple objects detectable per media object, via saliency detection / object localization.
+   No explicit user relevancy feedback.              | Accepting explicit user feedback signals.
+   No implicit user behavioral training.             | Training based on implicit user behavior signals.
+   Basic attack mitigation and spam control.         | Sophisticated attack mitigation.
+   Basic or no regression testing.                   | System to detect regressions caused by newly trained models.
+
+
+In summary, `V1 Indexer` will begin as a simple self-hosted system, composed of off-the-shelf open source components, having only a minimal wrapping JSON/REST API to provide abstraction, and using a small importer component for feeding data to and from the blockchain. This version will use simple, unsupervised models of perceptual similarity, based on simple image and textual heuristics.
+
+
+`Future Indexer` can be expanded to include features such as sophisticated neural CNN + attention models, training on supervised training corpuses setup to directly match the system's end-goals, comprehensive scoring and regression testing systems, and sophisticated handling of multi-modality mixed textual and media inputs. Last, a key challenge of `Future Indexer` will be in efficiently scaling the indexes over multiple machines in a P2P environment.
```

Comment:
this seems great. only comment i'd make is imo the indexer doesn't necessarily have to be p2p. maybe we could draw a spectrum of possible implementations, some of which are decentralized and others of which are simply distributed but centralized.